src/input/input_handler.cpp

```diff
@@ -0,0 +1,694 @@
+// SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "input_handler.h"
+
+#include "fstream"
+#include "iostream"
+#include "list"
+#include "map"
+#include "sstream"
+#include "string"
+#include "unordered_map"
+#include "vector"
+
+#include "SDL3/SDL_events.h"
+#include "SDL3/SDL_timer.h"
+
+#include "common/config.h"
+#include "common/elf_info.h"
+#include "common/io_file.h"
+#include "common/path_util.h"
+#include "common/version.h"
+#include "input/controller.h"
+
+namespace Input {
+/*
+Project structure:
+n to m connection between inputs and outputs
+Keyup and keydown events update a dynamic list* of u32 'flags' (what is currently in the list is
+'pressed') On every event, after flag updates, we check for every input binding -> controller output
+pair if all their flags are 'on' If not, disable; if so, enable them. For axes, we gather their data
+into a struct cumulatively from all inputs, then after we checked all of those, we update them all
+at once. Wheel inputs generate a timer that doesn't turn off their outputs automatically, but push a
+userevent to do so.
+
+What structs are needed?
+InputBinding(key1, key2, key3)
+ControllerOutput(button, axis) - we only need a const array of these, and one of the attr-s is
+always 0 BindingConnection(inputBinding (member), controllerOutput (ref to the array element))
+
+Things to always test before pushing like a dumbass:
+Button outputs
+Axis outputs
+Input hierarchy
+Multi key inputs
+Mouse to joystick
+Key toggle
+Joystick halfmode
+
+Don't be an idiot and test only the changed part expecting everything else to not be broken
+*/
+
+// Flags and values for varying purposes
+// todo: can we change these?
+int mouse_joystick_binding = 0;
+float mouse_deadzone_offset = 0.5, mouse_speed = 1, mouse_speed_offset = 0.1250;
+Uint32 mouse_polling_id = 0;
+bool mouse_enabled = false, leftjoystick_halfmode = false, rightjoystick_halfmode = false;
+
+std::list<std::pair<u32, bool>> pressed_keys;
+std::list<u32> toggled_keys;
+std::list<BindingConnection> connections = std::list<BindingConnection>();
```

Comment:
```suggestion
static std::vector<BindingConnection> connections;
```
I dont see any advantage to using a std::list, please correct I'm wrong